libs/parallelism/futures/include/hpx/futures/future.hpp

```diff
@@ -782,8 +781,91 @@ namespace hpx { namespace lcos {
         typedef detail::future_base<future<R>, R> base_type;
 
     public:
-        typedef R result_type;
-        typedef typename base_type::shared_state_type shared_state_type;
+        using result_type = R;
+        using shared_state_type = typename base_type::shared_state_type;
+
+        // Sender compatibility
+        template <template <typename...> class Tuple,
+            template <typename...> class Variant>
+        using value_types = std::conditional_t<std::is_void<result_type>::value,
+            Variant<Tuple<>>,
+            Variant<Tuple<std::add_rvalue_reference_t<result_type>>>>;
+
+        template <template <typename...> class Variant>
+        using error_types = Variant<std::exception_ptr>;
+
+        static constexpr bool sends_done = false;
+
+        template <typename Receiver>
+        struct operation_state
+        {
+            std::decay_t<Receiver> receiver;
+            std::decay_t<future> fut;
+
+            operation_state(operation_state&&) = delete;
+            operation_state(operation_state const&) = delete;
+
+            void start_helper(std::true_type) noexcept
+            {
+                try
+                {
+                    fut.then([receiver = std::move(receiver)](
```

Comment:
Would using `shared->set_on_completed()` be sufficient (it will avoid allocating a new future that is discarded anyways)?

Reply:
Yeah, that'd be a lot better. Will also change that.

Reply:
The only question I have is to on what execution context the continuation should execute. Do we have to do special things to enforce any possible constraints here?

Reply:
I'm glad you asked! That reminds me that I wanted to make the continuations `sync`, i.e. use whatever context the future is completed on (`set_on_completed` should have the same effect if I understand correctly). `execution::on` is there so that one can explicitly change context if needed.

Reply:
Yes, `set_on_completed` generally executes things `sync`, i.e. in the execution context of the thread that made the future ready.

Reply:
I changed this to use `set_on_completed`. I'd appreciate if you could have a look at it again.

Reply:
LGTM now, thanks!